src/Types.hs

```diff
@@ -0,0 +1,75 @@
+module Types where
+
+import Cardano.Prelude
+
+--import qualified Katip as Katip
+import qualified System.Metrics as Ekg
+
+-- | Cardano configuration
+data CardanoConfiguration = CardanoConfiguration
+    { scfgLogPath   :: !FilePath
+    , scfgDBPath    :: !FilePath
+    , scfgSomeParam :: !Int
+    }
+
+-- | The common runtime environment for all features in the server.
+-- All features have access to this environment.
+data CardanoEnvironment = CardanoEnvironment
+    { serverLogEnv      :: Text --Katip.LogEnv
+    , serverEkgStore    :: Ekg.Store
+     -- ...
+    }
+
+-- | Initalise 'ServerEnv'
+initializeCardanoEnvironment :: IO CardanoEnvironment
+initializeCardanoEnvironment = do
+  --  logenv   <- Katip.initLogEnv (...) (...)
+    ekgStore <- Ekg.newStore
+    return CardanoEnvironment {serverLogEnv = "To implement", serverEkgStore = ekgStore}
+
+loadCardanoConfiguration :: IO CardanoConfiguration
+loadCardanoConfiguration = do
+    pure $ CardanoConfiguration mempty mempty 0
+
+
+-- | The feature types we can have in the project.
+data FeatureType
+    = LoggingMonitoringFeature
+    | NetworkingFeature
+    | BlockchainFeature
+    | LedgerFeature
+    | WalletFeature
+    deriving (Eq, Show)
+
+-- | The option to not have any additional dependency for the @CardanoFeature@.
+data NoDependency = NoDependency
+    deriving (Eq, Show)
+
+-- | The option to not have any additional configuration for the @CardanoFeature@.
+data NoConfiguration = NoConfiguration
+    deriving (Eq, Show)
+
+-- | Cardano feature initialization.
+-- We are saying "you have the responsibility to make sure you use the right context!".
+data CardanoFeatureInit dependency configuration layer = CardanoFeatureInit
+    { featureType                   :: !FeatureType
+    -- ^ The type of the feature that we use.
+    , featureInit                   :: CardanoEnvironment -> dependency -> CardanoConfiguration -> configuration -> IO layer
+    -- ^ Again, we are not sure how is the user going to run the actual feature,
+    -- so we provide him with the most flexible/powerful context we have, @IO@.
+    -- Notice the arrangement of the parameters - specific, general, specific, general, result.
+    , featureCleanup                :: layer -> IO ()
```

Comment:
Do we need both a `featureCleanup` and `featureShutdown`?

Reply:
`featureShutdown` is for explicitly shutting down the feature (when the developers want to do that) and `featureCleanup` is when the developers want to clean something up before the feature is shut down - for example, saving some logs or closing files or something similar.

`featureShutdown` is under the control of the developers, while `featureCleanup` is a bracket cleanup call, basically.